editor/utils/heartbeat.js

```diff
@@ -0,0 +1,173 @@
+import store from '../store';
+import {
+	getCurrentPost,
+	getEditedPostTitle,
+	getEditedPostExcerpt,
+	getEditedPostContent,
+} from '../store/selectors';
+import {
+	toggleAutosave,
+} from '../store/actions';
+
+export function setupHeartbeat() {
+	const $document = jQuery( document );
+	/**
+	 * Configure heartbeat to refresh the wp-api nonce, keeping the editor authorization intact.
+	 */
+	$document.on( 'heartbeat-tick', ( event, response ) => {
+		if ( response[ 'rest-nonce' ] ) {
+			window.wpApiSettings.nonce = response[ 'rest-nonce' ];
+		}
+	} );
+
+	/**
+	 * Configure Heartbeat post locks.
+	 *
+	 * Used to lock editing of an object by only one user at a time.
+	 *
+	 * When the user does not send a heartbeat in a heartbeat-time
+	 * the user is no longer editing and another user can start editing.
+	 */
+	$document.on( 'heartbeat-send.refresh-lock', function( e, data ) {
+		const lock = jQuery( '#active_post_lock' ).val(),
+			postId = jQuery( '#post_ID' ).val(),
+			send = {};
+
+		if ( ! postId ) {
+			return;
+		}
+		send.post_id = postId;
+
+		if ( lock ) {
+			send.lock = lock;
+		}
+
+		data[ 'wp-refresh-post-lock' ] = send;
+	} );
+
+	// Refresh post locks: update the lock string or show the dialog if somebody has taken over editing.
+	$document.on( 'heartbeat-tick.refresh-lock', function( e, data ) {
+		if ( data[ 'wp-refresh-post-lock' ] ) {
+			const received = data[ 'wp-refresh-post-lock' ];
+
+			if ( received.lock_error ) {
+				// @todo suspend autosaving
+				// @todo Show "editing taken over" message.
+			} else if ( received.new_lock ) {
+				jQuery( '#active_post_lock' ).val( received.new_lock );
+			}
+		}
+	} );
+
+	/**
+	 * @summary Concatenates the title, content and excerpt.
+	 *
+	 * This is used to track changes when auto-saving.
+	 *
+	 * @since 3.9.0
+	 *
+	 * @param {Object} state The current state.
+	 *
+	 * @returns {string} A concatenated string with title, content and excerpt.
+	 */
+	const getCompareString = function( state ) {
+		return ( getEditedPostTitle( state ) || '' ) + '::' + ( getEditedPostContent( state ) || '' ) + '::' + ( getEditedPostExcerpt( state ) || '' );
```

Comment:
sure, thats much cleaner and easier to read, I'll update it. This function was copied from autosave.js and changed to use the editor state rather than reading from the dom.

Reply:
Updated in https://github.com/WordPress/gutenberg/pull/4218/commits/838f78fda82c326205a34a888f5c6d90889c2e88